src/settings/toml/migrations.rs

```diff
@@ -0,0 +1,504 @@
+            account_id,
+        );
+
+        let client = http::legacy_auth_client(user);
+        let res = client.get(&list_scripts_addr).send()?;
+        let scripts = res.json::<ListScriptsV4ApiResponse>()?.result;
+        let script = scripts.iter().find(|script| script.name == script_name);
+
+        match script {
+            None
+            | Some(ApiScript {
+                name: _,
+                migration_tag: None,
+            }) => Self::coalesce_migrations(&self.migrations, None),
+            Some(ApiScript {
+                name: _,
+                migration_tag: Some(tag),
+            }) => {
+                let mut iter = self
+                    .migrations
+                    .iter()
+                    .enumerate()
+                    .filter(|(_, m)| m.tag.as_ref() == Some(&tag));
+
+                match (iter.next(), iter.next()) {
+                    (Some((i, _)), None) if i + 1 < self.migrations.len() => {
+                        Self::coalesce_migrations(&self.migrations[i + 1..], Some(tag))
+                    }
+                    (Some(_), None) => Ok(ApiMigration {
+                        old_tag: Some(tag.clone()),
+                        new_tag: Some(tag.clone()),
+                        migration: Migration::default(),
+                    }),
+                    (Some(_), Some(_)) => failure::bail!(
+                        "Multiple migrations with tag {} were found in your wrangler.toml",
+                        tag
+                    ),
+                    (None, _) => failure::bail!(
+                        "The migration with tag {} was not found in your wrangler.toml.",
+                        tag
+                    ),
+                }
+            }
+        }
+    }
+
+    fn coalesce_migrations(
+        migrations: &[MigrationConfig],
+        old_tag: Option<&str>,
+    ) -> Result<ApiMigration, failure::Error> {
+        if migrations.is_empty() {
+            failure::bail!("cannot coalesce empty migrations");
+        }
+        log::info!(
+            "coalescing migrations {:#?} after {:?}",
+            migrations,
+            old_tag
+        );
+        let mut new_classes = migrations[0].migration.durable_objects.new_classes.to_vec();
+        let mut deleted_classes = migrations[0]
+            .migration
+            .durable_objects
+            .deleted_classes
+            .to_vec();
+        let mut renamed_classes = migrations[0]
+            .migration
+            .durable_objects
+            .renamed_classes
+            .to_vec();
+        let mut transferred_classes = migrations[0]
+            .migration
+            .durable_objects
+            .transferred_classes
+            .to_vec();
+
+        for mc in &migrations[1..] {
+            // Renames will modify existing new, rename, and transfer migrations if they are present
+            for rename in mc.migration.durable_objects.renamed_classes.iter() {
```

Comment:
If someone has multiple rename migrations away from the same class name (which is also a target of one of the prior migrations), then this will probably lead to weird error messages from the API (along the lines of "class doesn't exist"). But that should be quite uncommon, and it wouldn't be easy to do better here without a bunch more error checking that's already done in the API.

So this looks good 👍

Reply:
Although I suppose some of this logic could go awry if the initial set of migrations is invalid. For example, if there are multiple matches for a given classname, won't this code only find and update the first? It's not valid to have multiple migrations targeting the same class, but unless wrangler has already validated that then it isn't really safe to assume here.

I think we either need to do that validation or attempt to update all matching migrations in this logic rather than just the first one. I'd personally suggest doing the former rather than the latter, since it seems simpler and more self-contained. You could probably just create a `set`, and as you process each individual migration (including the first batch of them above this loop), make sure its destination class isn't already the destination of a different migration.